rewrite-core/src/main/java/org/openrewrite/scheduling/RecipeRunCycle.java

```diff
@@ -181,6 +181,7 @@ public LSS editSources(LSS sourceSet) {
                         return null;
                     }
 
+                    recipeRunStats.recordSourceVisited(source);
```

Comment:
Perhaps good to get input on the approach first before we continue, to make changes, but pointing out a detail:

By calling this method here we're executing this not once for every file, but for every recipe that handles a file (so again a multiplication). Perhaps it's better called not for every items on the recipeStack, but between lines 177 and 178 for every sourceFile processed there.